### PR TITLE
feature(opportunity-outside-range-c1-c2)

### PR DESCRIPTION
**Pull request by lukedawilson**
- Fixes to `ValidFromBeforeStartDate` generation logic
- Support for `TestOpportunityBookableOutsideValidFromBeforeStartDate`
- More linter cleanup

### CODE REVIEW

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 1]
```diff
@@ -903,12 +884,6 @@ public void DeleteTestFacilitiesFromDataset(string testDatasetIdentifier)
                     where: x => x.TestDatasetIdentifier == testDatasetIdentifier && !x.Deleted);
             }
         }
-
-        private static TimeSpan? ValidFromBeforeStartDate(Randomizer random)
-        {
-            return random.Int(1, 5) == 1 ? TimeSpan.FromDays(random.Int(1, 5)) : default(TimeSpan?); // set on 20% of opportunities
-        }
-
         private static TimeSpan? ValidFromBeforeStartDate(bool? validFromBeforeStartDate, DateTime startDate)
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:887]: Perhaps worth some comments in here explaining the resulting behavior?

**Review by nickevansuk — APPROVED** [review]
Looks good! Approved after one minor amend

In future suggest we break out major lint fixes to a separate PR as it makes it much easier to review, and helps with traceability too

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs (hunk 1 above)
Comment (nickevansuk) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:887]: Also is there a risk that this logic creates a negative validFromBeforeStartDate, if the event is happening in the next 2 hours?

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 2]
```diff
@@ -741,35 +719,38 @@ private static void CreateFakeFacilitiesAndSlots(IDbConnection db)
...
+                {
+                    Id = id,
+                    Deleted = false,
+                    Title = $"{Faker.Commerce.ProductMaterial()} {Faker.PickRandomParam("Yoga", "Zumba", "Walking", "Cycling", "Running", "Jumping")}",
+                    Price = decimal.Parse(Faker.Random.Bool() ? "0.00" : Faker.Commerce.Price(0, 20)),
+                    RequiresApproval = Faker.Random.Bool(),
+                    SellerId = Faker.Random.Long(1, 3),
+                    ValidFromBeforeStartDate = ValidFromBeforeStartDate(
+                        Faker.Random.Bool() ? Faker.Random.Bool() : (bool?)null,
+                        occurrences.Single(o => o.Id == id * 10).Start)
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:751]: Is there a more robust way of achieving this? Would be good to add a comment on what this logic does too

**Review by nickevansuk — CHANGES_REQUESTED** [review]
Sorry have spotted something else here

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs (hunk 2 above)
Comment (lukedawilson) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:751]: It's robust, as it mirrors the logic above. I could make it more efficient by using a dictionary. I wondered whether to bother, but now you've convinced me.

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs (hunk 1 above)
Comment (lukedawilson) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:887]: I think this is handled by:
```
case false when startDate.Date == now.Date:
                    return TimeSpan.Zero;
```

Unless I've missed something.

Comment (lukedawilson) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:888]: Changed it to return one hour.

**Review by nickevansuk — CHANGES_REQUESTED** [review]
Still some questions, see comments above

**Review by nickevansuk — CHANGES_REQUESTED** [review]
Comments from earlier conversation

**Pull request closed without merging**